### ISSUE TITLE
Annotationless polymorphic union types

### ISSUE THREAD

**Issue by Rurouni** [neuecc/ZeroFormatter#32]
Hi,

I wonder how hard it would be to support something similar to protobuf-net's RuntimeTypeModel.AddSubType() so that one can configure inheritance trees with some dynamic rules instead of static annotation + enum + override?
When modelling protocol with big open inheritance trees like Message or Event it is very convenient if you can use convention based approach (for example using reflection scan + consistent hash of message type name as id in union type, works well even with complicated Unity targets like IL2CPP if you can generate dynamic model in full .net runtime environment and then feed it into static protocol compiler to generate stripped protocol dll like in protobuf-net)

Thank you
Sergii

**Comment by neuecc** [comment]
Thank you.

Hmn, is this concept match for you?

```
public interface IDynamicRoot
{
}

[ZeroFormattable]
public class TypeA {}

[ZeroFormattable]
public class TypeB{}

// RegisterDynamic? : key as int(binary size)
Formatter.RegisterDynamicUnion<IDynamicRoot, int>(x =>
{
    x.Register(key: 0, subType: typeof(TypeA));
    x.Register(key: 1, subType: typeof(TypeB));
});

// RegisterDynamic? : key as string(convenient)
Formatter.RegisterDynamicUnion<IDynamicRoot, int>(x =>
{
    x.Register(key: typeof(TypeA).FullName, subType: typeof(TypeA));
    x.Register(key: nameof(TypeB).FullName, subType: typeof(TypeB));
});

// Serialize/Deserialize...
var bin = ZeroFormatterSerializer.Serialize<IDynamicRoot>(new TypeA());

// o is TypeA
var o = ZeroFormatterSerializer.Deserialize<IDynamicRoot>(bin);
```

**Comment by Rurouni** [comment]
Ideally that's what I was thinking about
``` CSharp
//consumer types
public class MessageBase {}
public interface IEvent{}

public class UnknownEvent : IEvent {}
[ZeroFormattable]
public class EventA : IEvent {}
[ZeroFormattable]
public class EventB : IEvent {}

public class UnknownMessage : MessageBase {}
[ZeroFormattable]
public class MessageA : MessageBase {}
[ZeroFormattable]
public class MessageB : MessageBase 
{
    [Index(0)]
    public virtual List<IEvent> Events { get; set; }
}

[ZeroFormattable]
public class MessageBatch : MessageBase 
{
    [Index(0)]
    public virtual List<MessageBase> Messages { get; set; }
}

//serialiser registration
var serialisationModel = ZeroFormatterSerializer.CreateDefaultModel()

serialisationModel.RegisterTypesWithAttributes(someAssembly);

serialisationModel.RegisterPolymorphicType<IEvent, int, UnknownEvent>(x =>
{
    //can be easily extended to reflection based scan if library consumer wants it
    x.Register(key: SomeIdGen(nameof(EventA)), subType: typeof(EventA));
    x.Register(key: SomeIdGen(nameof(EventB)), subType: typeof(EventB));
});

serialisationModel.RegisterPolymorphicType<MessageBase, int, UnknownMessage>(x =>
{
    x.Register(key: SomeIdGen(nameof(MessageA)), subType: typeof(MessageA));
    x.Register(key: SomeIdGen(nameof(MessageB)), subType: typeof(MessageB));
    x.Register(key: SomeIdGen(nameof(MessageBatch)), subType: typeof(MessageBatch));
});

MessageBase msg  = new MessageBatch { Messages = new List<MessageBase> { new MessageB { Events = new List<IEvent> { new EventA() }}};

var serialiser  = ZeroFormatterSerializer.CreateFromModel(serialisationModel);
var bts = serialiser.Serialise(msg);
var message = serialiser.Deserialise<MessageBase>(bts)

ZeroFormatterProtocolGenerator.CompileFromModel(serialisationModel, "path\UnityCompatibleProtocol.dll");
```
The reason for UnknownMessage/UnknownEvent is backward compatibility, so that when you read List<MessageBase> in older client instead of failing on deserialisation it can just fallback to set them as UnknownMessage allowing layer above to skip new unknown messages without losing known ones.

The reason for non-static serialiser class is so you can have differently configured serialisers in same AppDomain. It's generally better to have non-static implementation that is exposed in preconfigured static wrapper class, because you can always create a user friendly singleton from non-static class but not other way around. 

Also because protocol is generated at runtime we need to be able to generate precompiled dll for unity from it, so in addition to command line tool I suggest exposing protocol generator as part of library

I really like the library btw, if it wasn't for the protocol like above I would have already switched from Protobuf-Net to ZeroFormatter

**Comment by neuecc** [comment]
Thank you, I understood.
Based on your suggestion, I thought about a new API.

```csharp
public interface ITypeResolver
{
    // can be precompiled for improve initial performance(or AOT environment).
    // precompiled generator always set to false.
    bool IsUseDefaultDynamicSerializer { get; }

    // AppendFormatterResolver in normal.
    object ResolveFormatter(Type type);

    // RegisterPolymorphicType
    object ResolveDynamicUnion(Type unionType, DynamicUnionResolver resolver);
}

public class CustomSerializationModel : ITypeResolver
{
    public bool IsUseDefaultDynamicSerializer
    {
        get
        {
            return true;
        }
    }

    public object ResolveFormatter(Type t)
    {
        return null;
    }

    public void ResolveDynamicUnion(Type unionType, DynamicUnionResolver resolver)
    {
        if (unionType == typeof(IEvent))
        {
            // resolver api is [Type, object] for reflection based regsiter(but not typesafe)
            resolver.RegisterDiscriminatedKeyType(typeof(int));
            resolver.RegisterFallbackType(typeof(UnknownEvent));
            resolver.RegisterSubType(key: SomeIdGen(nameof(EventA)), subType: typeof(EventA));
        }
        // else if...
    }
}

// chnage Formatter<T> to here
    public abstract class Formatter<TTypeResolver, T>
        where TTypeResolver : ITypeResolver, new()
```

```csharp
// usage

// mark DynamicUnion on base type...
[DynamicUnion]
public class MessageBase {}
[DynamicUnion]
public interface IEvent{}

/// use dynamic serializer
ZeroFormatterSerializer.CustomSerializer<CustomSerializationModel>.Serialize();

// create dll
// ZeroFormatterProtocolGenerator.CompileDefault();
// ZeroFormatterProtocolGenerator.Compile<CustomSerializationModel>(IEnumerable<Type> generateTypes,string typeFullName, string path);
// ZeroFormatterProtocolGenerator.Compile<CustomSerializationModel>(IEnumerable<Assembly> scanAssemblies,string typeFullName, string path);

// use dll
// ZeroFormatterSerializer.SetDefault<CustomSerializationModel>();
// ZeroFormatterSerializer.Serialize();
```

Serializer acquisition by static type is the core of ZeroFormatter and can not be changed.
However, it can be solved by static type with custom type.

I hope these changes will satisfy your requirements, and I want you to switch:)

**Comment by Rurouni** [comment]
That looks good.
My only question is around performance with `CustomSerializationModel`... to make it fast reflection based scan should happen only once and mapping of type->formatter should be cached otherwise `ZeroFormatterSerializer.CustomSerializer<CustomSerializationModel>().Serialize();` will be prohibitively expensive. 
Or is you plan that consumer just saves instance of serialiser : 
``` CSharp
//once on startup
var serialiser = ZeroFormatterSerializer.CustomSerializer<CustomSerializationModel>();

//many times, threadsafe
serialiser.Serialise();
```
Cheers

**Comment by neuecc** [comment]
`ZeroFormatterSerializer.CustomSerializer<CustomSerializationModel>.Serialize<T>(T value)`
calls 
`Formatter<CustomSerializationModel, T>.Default.Serialize(T value)`;
so
`Formatter<CustomSerializationModel, T>.Default` is automatically cached on static field.

**Comment by Rurouni** [comment]
Understood, but how the caching for `void ResolveDynamicUnion(Type unionType, DynamicUnionResolver resolver)` will be working? As read through cache on first time you serialise something that contains IEvent(for example)?

**Comment by neuecc** [comment]
Yes, that cache will be done for the first time.
This is the flow at the time of caching.

`CustomSerializer<CustomSerializationModel>.Serialize<IEvent>()`
->
`Formatter<CustomSerializationModel, IEvent>.Default`
->
`ResolveDynamicUnion(typoef(IEvent), resolver)`
->
`Formatter<CustomSerializationModel, IEvent>.Default = resolver.CreateFormatter()`

**Comment by neuecc** [comment]
Today I released `1.5.0` and it supports `DynamicUnion`.
Currently supports `ITypeResolver configuraiton`, `RegisterDynamic`, `Fallback type on Union`.
DLL code generate is not yet.
